src/lib/InstantSearch.js

```diff
@@ -296,87 +236,70 @@ See: https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-o
       );
       this._createURL = routingManager.createURL.bind(routingManager);
       this._createAbsoluteURL = this._createURL;
-      this.widgets.push(routingManager);
+      // We don't use `addWidgets` because we have to ensure that `RoutingManager`
+      // is the last widget added. Otherwise we have an issue with the `routing`.
+      // https://github.com/algolia/instantsearch.js/pull/3149
+      this.mainIndex.getWidgets().push(routingManager);
     } else {
       this._createURL = defaultCreateURL;
       this._createAbsoluteURL = defaultCreateURL;
       this._onHistoryChange = noop;
     }
 
-    const initialSearchParameters = this.widgets.reduce(
-      enhanceConfiguration,
-      this._searchParameters
-    );
-
-    const helper = algoliasearchHelper(
-      this.client,
-      initialSearchParameters.index || this.indexName,
-      initialSearchParameters
-    );
-
-    const derivedHelper = helper.derive(() => helper.state);
+    // This Helper is used for the queries, we don't care about its state. The
+    // states are managed at the `index` level. We use this Helper to create
+    // DerivedHelper scoped into the `index` widgets.
+    const mainHelper = algoliasearchHelper(this.client);
 
-    helper.search = () => {
+    mainHelper.search = () => {
       // This solution allows us to keep the exact same API for the users but
       // under the hood, we have a different implementation. It should be
       // completely transparent for the rest of the codebase. Only this module
       // is impacted.
-      helper.searchOnlyWithDerivedHelpers();
+      return mainHelper.searchOnlyWithDerivedHelpers();
```

Comment:
seems like this was from a previous PR, but what about:

```
mainHelper.search = mainHelper.searchOnlyWithDerivedHelpers
```

Reply:
Both are equivalent.